broker.go

```diff
@@ -0,0 +1,706 @@
+
+	mu       sync.Mutex
+	metadata clusterMetadata
+	conns    map[int32]*connection
+}
+
+// Dial connects to any node from given list of kafka addresses and after
+// successful metadata fetch, returns broker.
+// Returned broker is not initially connected to any kafka node.
+func Dial(nodeAddresses []string, conf BrokerConf) (*Broker, error) {
+	broker := &Broker{
+		conf:  conf,
+		conns: make(map[int32]*connection),
+	}
+
+	for _, addr := range nodeAddresses {
+		conn, err := NewConnection(addr, conf.DialTimeout)
+		if err != nil {
+			conf.Log.Printf("could not connect to %s: %s", addr, err)
+			continue
+		}
+		defer func(c *connection) {
+			_ = conn.Close()
+		}(conn)
+		resp, err := conn.Metadata(&proto.MetadataReq{
+			ClientID: broker.conf.ClientID,
+			Topics:   nil,
+		})
+		if err != nil {
+			conf.Log.Printf("could not fetch metadata from %s: %s", addr, err)
+			continue
+		}
+		broker.rewriteMetadata(resp)
+		return broker, nil
+	}
+	return nil, errors.New("could not connect")
+}
+
+// Close is closing broker and all active kafka nodes connections.
+func (b *Broker) Close() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	for nodeID, conn := range b.conns {
+		if err := conn.Close(); err != nil {
+			b.conf.Log.Printf("failed closing node %d connection: %s", nodeID, err)
+		}
+	}
+}
+
+func (b *Broker) Metadata() (*proto.MetadataResp, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.fetchMetadata()
+}
+
+// refreshMetadata is requesting metadata information from any node and refresh
+// internal cached representation.
+// Because it's changing internal state, this method requires lock protection,
+// but it does not acquire nor release lock itself.
+func (b *Broker) refreshMetadata() error {
+	meta, err := b.fetchMetadata()
+	if err == nil {
+		b.rewriteMetadata(meta)
+	}
+	return err
+}
+
+// fetchMetadata is requesting metadata information from any node and return
+// protocol response if successful
+// Because it's using metadata information to find node connections it's not
+// thread safe and using it require locking.
+func (b *Broker) fetchMetadata() (*proto.MetadataResp, error) {
+	checkednodes := make(map[int32]bool)
+
+	// try all existing connections first
+	for nodeID, conn := range b.conns {
+		checkednodes[nodeID] = true
+		resp, err := conn.Metadata(&proto.MetadataReq{
+			ClientID: b.conf.ClientID,
+			Topics:   nil,
+		})
+		if err != nil {
+			b.conf.Log.Printf("cannot fetch metadata from node %d: %s", nodeID, err)
+			continue
+		}
+		return resp, nil
+	}
+
+	// try all nodes that we know of that we're not connected to
+	for nodeID, addr := range b.metadata.nodes {
+		if _, ok := checkednodes[nodeID]; ok {
+			continue
+		}
+		conn, err := NewConnection(addr, b.conf.DialTimeout)
+		if err != nil {
+			b.conf.Log.Printf("could not connect to %s: %s", addr, err)
+			continue
+		}
+		// we had no active connection to this node, so most likely we don't need it
+		defer func(c *connection) {
+			_ = c.Close()
+		}(conn)
```

Comment:
This defer captures a variable which it doesn't use. Also, why wrap this in a closure anyway?